Add properties to MakeDeps generator

vajdaz opened the pull request:
Changelog: Feature: MakeDeps generator generates make variables for dependencies and their components.
Docs: https://github.com/conan-io/docs/pull/3794

Fixes #16572

vajdaz committed:
Add properties to MakeDeps generator

If the receipt defines custom properties in the package_info() method
then these properties should appear as make variables, too. The pattern
of the make variable is

CONAN_PROPERTY_{dependency_name}_{property_name}

vajdaz committed:
Add property feature for components

Furthermore handle case when _properties is None.

vajdaz committed:
Extracting property makefication

vajdaz committed:
Add some tests

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/conan-io/conan?pullRequest=16613) <br/>All committers have signed the CLA.

memsharded commented on test/integration/toolchains/gnu/test_makedeps.py:
```diff
@@ -44,6 +45,7 @@ def package_info(self):
     assert f'CONAN_LIB_DIRS_MYLIB = \\\n\t$(CONAN_LIB_DIR_FLAG){prefix}/my_absoulte_path/fake/mylib/lib \\\n\t$(CONAN_LIB_DIR_FLAG)$(CONAN_ROOT_MYLIB)/lib2' in makefile_content
     assert f'CONAN_INCLUDE_DIRS_MYLIB = $(CONAN_INCLUDE_DIR_FLAG){prefix}/my_absoulte_path/fake/mylib/include' in makefile_content
     assert 'CONAN_BIN_DIRS_MYLIB = $(CONAN_BIN_DIR_FLAG)$(CONAN_ROOT_MYLIB)/bin' in makefile_content
+    assert 'CONAN_PROPERTY_MYLIB_MY_PROP = my prop value' in makefile_content
```
How does Make deal with spaces? Should this be quoted?

vajdaz replied:
This is a ticky question. In GNU Make the value of a variable is a list of wiords separated by whitespaces. The number of whitespaces has no relevance. So following lines are kind of the same:
```
XXX = these are four words

XXX = these    are  four          words
```
Quotes would be part of the word. You can escape spaces with backslash.
```
XXX = first\ word second\ word
``` 
Here `XXX` is a list of two words, one is "first word" and the second is "second word".

The question is: what does the user want to achieve if he/she assigns a property the value "these are four words". You could escape spaces with backslashes. You could leave it as is. What is the usage on the receiver side? I would leave the decision in the hands of the person who sets the property.

And if we do sommething with spaces, what to do with other special characters like newlines? Values with newlines could be mapped to following makefile syntax:

```
define MULTILINE_MAKE_VAR
value
over
multiple
lines
endef
```

Are these real use cases? Will somebody add a property with newlines? Should somebody add a property with newlines? I grepped though conan-center-iindex and found no properties with newlines.

However, I found some that contain paths, which potentionally could contain spaces, if the conan cache directory or an editable directory path contains spaces.

So maybe escaping spaces should be done after all. But then look at this makefile:

```
XXX = aaa bbb
YYY = aaa\ bbb

.PHONY: aaa bbb aaa\ bbb

all: use_case_1 use_case_2 use_case_3

use_case_1:
	@echo "$(XXX)"
	@echo "$(YYY)"

use_case_2: $(XXX)
	@echo $<

use_case_3: $(YYY)
	@echo $<
```

The output of `make` is:

```
aaa bbb
aaa\ bbb
aaa
aaa bbb
```

So if you will use the value in a receipt, escaping is not a good idea. If you will use it as a prerequisite, it is a good idea. But what will be the use case? That depends on your make system integration.

Based on my expericence with my own make based build system integration I assume that the variables will land in a receipe (e.g. as compiler/linker options). Therefore I vote on not escaping whitespaces.

What are your thoughts?

vajdaz replied:
Spaces in paths and `MakeDeps` seems to be a deep rabbit hole. I tested what MakeDeps generates if I make a package editable and have a space in the editable path (even without my changes). I do following in the shell:

```
conan editable add ~/my/path\ with/space 
``` 
MakeDeps wil create a makefile like this when I install the package:
```
CONAN_ROOT_MYLIB = /home/myuser/my/path with/space

CONAN_INCLUDE_DIRS_MYLIB = \
	$(CONAN_INCLUDE_DIR_FLAG)$(CONAN_ROOT_MYLIB)/include \
	$(CONAN_INCLUDE_DIR_FLAG)$(CONAN_ROOT_MYLIB)/h
```
Using `CONAN_INCLUDE_DIRS_MYLIB` in a receipt to pass the include folders to the compiler will not work in this situation. I assume that you don't receive bug reports about this, because no sane developer name their folders like this.

vajdaz replied:
Today I found this:

https://savannah.gnu.org/bugs/?712

Because of its weird space handling GNU Make cannot handle paths with spaces correctly, and there seems to be no easy solution for that.

This makes me also prefering not to escape spaces.

memsharded replied:
> Based on my expericence with my own make based build system integration I assume that the variables will land in a receipe (e.g. as compiler/linker options). Therefore I vote on not escaping whitespaces.

Yes, I think I agree, if there is no really better default, the best is not doing or manipulating the inputs at all.

> Because of its weird space handling GNU Make cannot handle paths with spaces correctly, and there seems to be no easy solution for that.

That is true, I recall that we also have a ``TestClient(path_with_spaces=False)`` to allow for tests that won't support in any way paths with spaces, and Make/Autotools tests are common reasons.


Lets not do anything with spaces, and add some comment in the docs that this will be problematic and something on the user side.

vajdaz replied:
Ok. During finalization of the docs a last important thought came into my mind. A newline in the property value may break the generated makefile syntax. A makefile with broken syntax is worse than a make variable with wrong value. Therefore, maybe the generoator should convert newlines in spaces.

The alternative would be using the `define`, `endef`syntax if the value contains newlines. However, even that does not prevent the makefile from breaking. As an example:

```
define XXX
two
lines
endef

mytarget:
	echo "$(XXX)"
```

This will not work, because it expands to

```
define XXX
two
lines
endef

mytarget:
	echo "two
lines"
```

But

```
define XXX
echo two
echo lines
endef

mytarget:
	$(XXX)
```

works. And that's why I would convert a multiline value to one line. I cannot imagine any use case that would need multilne properties.

memsharded replied:
There is another approach, it is also possible to:
- Do a warning and do nothing, recommend the users to file a ticket and give feedback (it is more likely that users will realize about the property not being defined than having a modified value)
- Do the same but with a hard error saying it is not supported at the moment

vajdaz replied:
I like the policy of "don't break userspace". When using MakeDeps, a hard error would break existing conanfile.py receipts that worked before with this generator. A warning may be a hint for the package publisher but it would not prevent existing custom makefile based build system integrations from breaking if they use packages with receipts that set properties with newlines. There should be no such receipts out there in the wild, but I have the experience, that if something is possible, it is out there somewhere. The only solution to not break any kind of existing integration is to change newlines into spaces or (another new alternative) to give a warning and skip the property that has a newline. So they will not appear in the generated makefile.

memsharded replied:
> I like the policy of "don't break userspace".

I agree that this sounds better for this specific use case.

As the ``CONAN_PROPERTY_`` are not defined yet, not adding one would be perfectly valid, and I think modifying some user property value shouldn't be done, if the value contains multiline, removing them will probably also break, but unexpectedly because the user might use the ``CONAN_PROPERTY_`` that will exist, but with a modified value. 

Then I suggest:
- Do a warning, not an error, so it doesn't break
- Do not generate the ``CONAN_PROPERTY_``

vajdaz replied:
Done. Docs updated, too.

memsharded commented on conan/tools/gnu/makedeps.py:
```diff
@@ -67,6 +68,15 @@ def _makefy(name: str) -> str:
     return re.sub(r'[^0-9A-Z_]', '_', name.upper())
 
 
+def _makefy_properties(properties: dict) -> Optional[dict]:
```
Maybe you can simplify the Optional by:
- Always return a dict, an empty one instead of None
- That also simplifies the template, that does no longer need to do ``{%- if values is not none -%}``

vajdaz replied:
done

memsharded reviewed (commented):
Looks good, thanks for the contribution @vajdaz !

vajdaz committed:
Review finding

- _makefy_properties() always returns a dict
- Simplification of define_multiple_variable_value jinja2 macro

memsharded commented on conan/tools/gnu/makedeps.py:
```diff
@@ -67,6 +68,15 @@ def _makefy(name: str) -> str:
+    :param properties: The property dictionary to be converted (None is also accepted)
+    :return: Modified property dictionary with keys not including bad characters that are not parsed correctly
+    """
+    return {_makefy(name): value for name, value in  properties.items()} if properties else {}
```
```suggestion
    return {_makefy(name): value for name, value in properties.items()} if properties else {}
```

memsharded reviewed (approved):
Looks good I think it can be moved out of draft and merged.
Many thanks!

vajdaz committed:
Remove unnecessary extra space

Co-authored-by: James <memsharded@gmail.com>

vajdaz committed:
Skip properties with newlines

It would break the makefile if we would create property values with
newlines. Therefore we output a warning message and skip such
properties.

memsharded commented on conan/tools/gnu/makedeps.py:
```diff
@@ -651,7 +670,7 @@ def generate(self) -> None:
                 continue
 
-            dep_gen = DepGenerator(dep, require)
+            dep_gen = DepGenerator(dep, require, self._conanfile.output)
```
```suggestion
            output = ConanOutput(scope=f"{self._conanfile} MakeDeps: {dep}:")
            dep_gen = DepGenerator(dep, require, output)
```
I'd say the warnings might be a bit limited in the information scope, not showing it is a warning from ``MakeDeps`` generator, and about one specific dep property.

It might be necessary to add an ``import``, this is mostly a rough idea.

vajdaz replied:
Done.

memsharded reviewed (approved):
Thanks!

vajdaz committed:
Display scope of warning

vajdaz commented:
Hi @memsharded,
The text in the release notes for 2.6.0 is a bit misleading. It says:

> Feature: MakeDeps generator generates make variables for dependencies and their components. (https://github.com/conan-io/conan/pull/16613)

However, the feautre is that the generator generates make variables for the *properties* of the direct dependencies and their components.

memsharded commented:
You are right. The release notes are generated automatically from the PR ``Changelog`` comment in https://github.com/conan-io/conan/pull/16613. If you think it is worth to change the changelog, a PR to the docs can be done, but it is probably not a big issue.
